### ISSUE TITLE
Improve error messages

### ISSUE THREAD

**Issue by killme2008** [GreptimeTeam/greptimedb#4340]
### What type of enhancement is this?

API improvement

### What does the enhancement do?

The error messages in GreptimeDB have some issues:

* Errors from DataFusion always appear as internal errors, which can be confusing and unhelpful for users.
* The MySQL protocol consistently returns an internal error type: https://github.com/GreptimeTeam/greptimedb/blob/da0c8402619134d455242bd5c16e16a3b3955994/src/servers/src/mysql/writer.rs#L233
* A similar problem exists in PostgreSQL: https://github.com/GreptimeTeam/greptimedb/blob/da0c8402619134d455242bd5c16e16a3b3955994/src/servers/src/postgres/handler.rs#L99
* Errors around timeout #3800 
* Some errors are mistakenly categorized as `Internal` or `Unknown`, causing the error message to not display correctly.
* If you have any other issues that you are aware of, please feel free to comment on them.
An effective system should offer clear information when errors occur.

### Implementation challenges

* [x] Correct all the error types that are wrongly categorized into `Internal` or `Unknown`.
* [ ] Returns the correct errors from DataFusion.
   * Proposed by @shuiyisong 
```
Add a special tag in front of the DataFusion error string, like "__gterror:",  
then serialize the original error information into it; if we detect this type of error message outside, 
deserialize it back to our original error.
```
* [x] Write the proper error kind for MySQL.
    * Mapping `StatusCode` to `ErrorKind`
    * https://greptimedb.rs/common_error/status_code/enum.StatusCode.html
    * https://docs.rs/opensrv-mysql/latest/opensrv_mysql/enum.ErrorKind.html
* [x] Write the proper error code for PostgreSQL.
    * Mapping `StatusCode` to PostgreSQL error codes
    * https://greptimedb.rs/common_error/status_code/enum.StatusCode.html
    * https://www.postgresql.org/docs/current/errcodes-appendix.html

_No response_

**Comment by leaf-potato** [comment]
I want to have a try for PostgreSQL.

**Comment by killme2008** [comment]
> I want to have a try for PostgreSQL.

Cool, thank you.

**Comment by leaf-potato** [comment]
@killme2008 I have a problem. Different from mysql crate, no type of PostgreSQL error codes found in pgwire crate. So in order to map StatusCode, I need to define enum PgWireErrorCode first? 
```rust
enum PgWireErrorCode {
    EC_00000,
    EC_01000,
    ...
}
```

**Comment by killme2008** [comment]
> @killme2008 I have a problem. Different from mysql crate, no type of PostgreSQL error codes found in pgwire crate. So in order to map StatusCode, I need to define enum PgWireErrorCode first?
> 
> ```rust
> enum PgWireErrorCode {
>     EC_00000,
>     EC_01000,
>     ...
> }
> ```

Yes, but i think you don't need to define all the pg error codes, but just the StatusCode needs.

https://greptimedb.rs/common_error/status_code/enum.StatusCode.html

**Comment by killme2008** [comment]
Or maybe a function like `fn status_code_to_pg_error(status: StatusCode) -> &str` is enough.
